File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.cpp
class Solution {
public:
    
    bool BFS(unordered_map<int, vector<int>>&mp, int st, int e, vector<bool>&visited){
        
        queue<int>pendingNodes;
        
        pendingNodes.push(st);
        
        visited[st] = true;
        
        while(!pendingNodes.empty()){
            
            int front = pendingNodes.front();
            
            pendingNodes.pop();
            
            if(front == e)     return true;
            
            for(int i : mp[front]){
                
                if(!visited[i]){
                    
                    pendingNodes.push(i);
                    
                    visited[i] = true;
                    
                }
                
            }
            
        }
        
        return false;
        
    }
    
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        
        // create graph         
        unordered_map<int, vector<int>>mp;
        
        for(auto &edge : edges){
            
            int u = edge[0];
            int v = edge[1];
            
            mp[u].push_back(v);
            mp[v].push_back(u);
            
        }
        
        //  create a visited array
        vector<bool>visited(n, false);
        
        return BFS(mp, source, destination, visited);
    }
};

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.java
class Solution {
    public boolean dfs(Map<Integer, List<Integer>> graph, boolean[] vis, int u, int dest){
        if(u == dest)     return true;
        
        vis[u] = true;
        
        for(int v : graph.get(u)){ 
            if(!vis[v])
                 if(dfs(graph, vis, v, dest))    return true;
        }
        
        return false;
    }
    
    public boolean validPath(int n, int[][] edges, int src, int dest) {
        // build graph
        Map<Integer, List<Integer>> graph = new HashMap<>();
        
        for(int i = 0; i < edges.length; i++){
            int u = edges[i][0];
            int v = edges[i][1];
            
            graph.putIfAbsent(u, new ArrayList<>());
            graph.putIfAbsent(v, new ArrayList<>());
            graph.get(u).add(v);
            graph.get(v).add(u);
        }
        
        boolean[] vis = new boolean[n];
        
        return dfs(graph, vis, src, dest);
    }
}